features/custom_codecs/README.md

```diff
@@ -0,0 +1,379 @@
+
+    public Address(String street, int zipcode) {
+        this.street = street;
+        this.zipcode = zipcode;
+    }
+
+    // getters, setters, equals() and hashcode() omitted
+}
+```
+
+First of all, create a codec for this class:
+
+```java
+public class AddressCodec extends TypeCodec.MappingCodec<Address, UDTValue> {
+
+    private final UserType userType;
+
+    public AddressCodec(TypeCodec<UDTValue> innerCodec, Class<Address> javaType) {
+        super(innerCodec, javaType);
+        userType = (UserType) innerCodec.getCqlType();
+    }
+
+    @Override
+    protected Address deserialize(UDTValue value) {
+        return value == null ? null : new Address(value.getString("street"), value.getInt("zipcode"));
+    }
+
+    @Override
+    protected UDTValue serialize(Address address) {
+        return address == null ? null : userType.newValue().setString("street", address.getStreet()).setInt("zipcode", address.getZipcode());
+    }
+}
+```
+
+Again, the driver provides a convenient base class: [MappingCodec][MappingCodec].
+
+Now, create your `CodecRegistry` and `Cluster` instances the usual way:
+
+```java
+CodecRegistry codecRegistry = new CodecRegistry();
+cluster = Cluster.builder()
+    .addContactPoints(...)
+    .withCodecRegistry(codecRegistry)
+    .build();
+```
+
+Now, retrieve the metadata for type `address`:
+
+```java
+UserType addressType = cluster.getMetadata().getKeyspace(...).getUserType("address");
+```
+
+And finally, register your `AddressCodec` with the `CodecRegistry`. Note that
+your codec simply delegates the hard work to an inner codec that knows how to deal with
+user type values:
+
+```java
+AddressCodec addressCodec = new AddressCodec(new TypeCodec.UDTCodec(addressType), Address.class);
+codecRegistry.register(addressCodec);
+```
+
+And voilà! You can now retrieve `address` values as `Address` instances:
+
+```java
+ResultSet rows = session.execute(...);
+Row row = rows.one();
+// let the driver convert it for you...
+Address address = row.get("address", Address.class);
+// ...or access the low-level UDTValue, if you need
+UDTValue addressUdt = row.get("address", UDTValue.class); // row.getUDTValue("address") would have worked too
+```
+
+### Limitations
+
+#### Generics
+
+Java generic types are fully supported, through Guava's [TypeToken][TypeToken] API. 
+Be sure to only use `get()` and `set()` methods
+that take a `TypeToken` argument:
+
+```java
+// this works
+Foo<Bar> foo = row.get(0, new TypeToken<Foo<Bar>>(){})
+// this does not work due to type erasure
+Foo<Bar> foo = row.get(0, Foo.class);
+```
+
+#### Type inheritance
+
+`CodecRegistry` looks for codecs that match exactly a given pair of `[CQL <-> Java]` types.
+
+As a consequence, type inheritance and interfaces are not supported. In other words, if you register a codec for 
+class `B extends A` and later require a conversion to class `A`,
+the driver will throw `CodecNotFoundException`, even if `B` instances are assignable to `A`:
```

Comment:
I had to re-read this a couple of times (even though the code example is clear). How about reformulating it this way:

As a consequence, type inheritance and interfaces are not supported. In other words, if `B` extends `A` and you register a codec for `B`, you can't pass `A` as the target type: 
```java
codecRegistry.register(new BCodec());
// this throws CodecNotFoundException, because no codec matches A.class exactly
A a = row.get(0, A.class);
// but this works
A a = row.get(0, B.class);
```